File: haskell/src/Utils/Input.hs
module Utils.Input
    ( getInput
    , getSolutionFromTxt
    , getSolutionFromTex
    ) where
import Text.Printf
import System.Directory
import System.Environment
import System.FilePath(joinPath)

getFile :: FilePath -> Int -> IO String
getFile ext problemNo = do
    execPath <- getExecutablePath
    readFile (joinPath [execPath, "..", "..", "..", "..", "..", "..", ext, printf "Euler%03d.%s" problemNo ext])

getInput :: Int -> IO String
getInput = getFile "in"

-- Convention for txt files: last line of text file contains solution to the problem.
getSolutionFromTxt :: Int -> IO String
getSolutionFromTxt problemNo =
    putStrLn "Reading solution from text file" >>
    last . filter (not . null) . lines <$> getFile "txt" problemNo

-- Convention for tex files: document ends like this:
-- \vspace{5ex}
-- \textbf{Answer:}
-- XYZ
-- \end{document}
-- Unless tex file is only explaining computations done.
getSolutionFromTex :: Int -> IO String
getSolutionFromTex problemNo =
    putStrLn "Reading solution from tex file" >>
    last . init . filter (not . null) . lines <$> getFile "tex" problemNo

File: clojure/src/project_euler/utils/input.clj
(ns project-euler.utils.input)

(defn get-file
  [ext no]
  (slurp (clojure.java.io/file ".." ext (format "Euler%03d.%s" no ext))))

(defn get-input
  [no]
  (get-file "in" no))

(defn get-txt-solution
  [no]
  (last (clojure.string/split-lines (get-file "txt" no))))

(defn get-tex-solution
  [no]
  (second (reverse (clojure.string/split-lines (get-file "txt" no)))))